signer/fulcio/fulcio.go

```diff
@@ -15,133 +15,286 @@
+		raw = tokenOption
 
-	if block.Type != "CERTIFICATE" {
-		return nil, fmt.Errorf("failed to parse certificate PEM")
+	case tokenOption == "" && isatty.IsTerminal(os.Stdin.Fd()):
+		tok, err := oauthflow.OIDConnect(oidcIssuer, oidcClientID, "", "", oauthflow.DefaultIDTokenGetter)
+		if err != nil {
+			return nil, err
+		}
+		raw = tok.RawString
+	default:
+		return nil, errors.New("no token provided")
 	}
 
-	leaf, err := x509.ParseCertificate(block.Bytes)
+	certResp, err := GetCert(ctx, key, fClient, raw)
 	if err != nil {
 		return nil, err
 	}
 
-	intermediates := []*x509.Certificate{}
-	roots := []*x509.Certificate{}
+	var chain *fulciopb.CertificateChain
 
-	rest := certResp.ChainPEM
+	switch cert := certResp.Certificate.(type) {
+	case *fulciopb.SigningCertificate_SignedCertificateDetachedSct:
+		chain = cert.SignedCertificateDetachedSct.GetChain()
+	case *fulciopb.SigningCertificate_SignedCertificateEmbeddedSct:
+		chain = cert.SignedCertificateEmbeddedSct.GetChain()
+	}
 
-	for len(rest) > 0 {
-		var block *pem.Block
-		block, rest = pem.Decode(rest)
-		if block == nil {
-			return nil, fmt.Errorf("failed to parse certificate PEM")
-		}
+	certs := chain.Certificates
+
+	var rootCACert *x509.Certificate
+	var intermediateCerts []*x509.Certificate
+	var leafCert *x509.Certificate
 
-		if block.Type != "CERTIFICATE" {
-			return nil, fmt.Errorf("failed to parse certificate PEM")
+	for _, certPEM := range certs {
+		certDER, _ := pem.Decode([]byte(certPEM))
+		if certDER == nil {
+			return nil, errors.New("failed to decode PEM block containing the certificate")
 		}
 
-		cert, err := x509.ParseCertificate(block.Bytes)
+		cert, err := x509.ParseCertificate(certDER.Bytes)
 		if err != nil {
-			return nil, err
+			return nil, fmt.Errorf("failed to parse certificate: %w", err)
 		}
 
-		switch cert.IsCA {
-		case true:
-			roots = append(roots, cert)
-		default:
-			intermediates = append(intermediates, cert)
+		if cert.IsCA {
+			if cert.Subject.CommonName == cert.Issuer.CommonName {
+				rootCACert = cert
+			} else {
+				intermediateCerts = append(intermediateCerts, cert)
+			}
+		} else {
+			leafCert = cert
 		}
-
 	}
 
 	ss := cryptoutil.NewRSASigner(key, crypto.SHA256)
-	if err != nil {
-		return nil, err
+	if ss == nil {
+		return nil, errors.New("failed to create RSA signer")
 	}
 
-	witnessSigner, err := cryptoutil.NewX509Signer(ss, leaf, intermediates, roots)
+	witnessSigner, err := cryptoutil.NewX509Signer(ss, leafCert, intermediateCerts, []*x509.Certificate{rootCACert})
 	if err != nil {
 		return nil, err
 	}
+
 	return witnessSigner, nil
 }
 
-func getCert(signer *signature.RSAPKCS1v15Signer, fc api.Client, oidcIssuer string, oidcClientID string) (*api.CertificateResponse, error) {
-	tok, err := oauthflow.OIDConnect(oidcIssuer, oidcClientID, "", "", oauthflow.DefaultIDTokenGetter)
+func GetCert(ctx context.Context, key *rsa.PrivateKey, fc fulciopb.CAClient, token string) (*fulciopb.SigningCertificate, error) {
```

Comment:
Is GetCert called from outside this package?  If not I think we should leave it unexported

Reply:
good point